### DISCUSSION TITLE
Joining values in "leaves"

### DISCUSSION THREAD

**Discussion by Limess** [cue-lang/cue#1652] [Q&A]
Hello!  Cross posting this from the #help Slack channel:

The following can be seen in a minimal reproduction repo here https://github.com/limess/cue-leaves-issue via:

```shell
cd $(mktemp -d)
(
set -e
git clone https://github.com/Limess/cue-leaves-issue
cd cue-leaves-issue
cue eval ./config/pipelines/... -e _monitored_pipelines_label
)
```

## Description

We just started working with cue to generate some Prometheus configuration.
I’ve hit a little bit of a conceptual issue which I was hoping someone could cleanup:
The project looks like this:

```
❯ tree
.
├── README.md
├── cue.mod
│   └── module.cue
└── config
    ├── common.cue
    ├── output_tool.cue
    ├── pipelines
    │   ├── article-pipeline
    │   │   └── pipeline.cue
    │   ├── pipeline.cue
    │   ├── pipeline_rules.cue
    │   ├── pipeline_slos.cue
    │   └── twitter-pipeline
    │       └── pipeline.cue
    ├── prometheus.cue
    └── sloth.cue
```

with the schemas in the root, and the data in the leaves (“pipeline”) definitions.
This works great for generating rules into separate files, however when trying to generate a configuration with labels based on a combination of both definitions of “pipeline” in the leaves, each leaf results in a separate definition:

```
_monitored_pipelines_label: "(" + strings.Join([ for k, v in pipeline {v.name}], "|") + ")"
```

results in:

```
❯ cue eval ./config/pipelines/... -e _monitored_pipelines_label
"()"
// ---
"(article-pipeline)"
// ---
"(twitter-pipeline)"
```

I feel I’m conceptually misunderstanding something here, it looks like each leaf, and then also something else ([]) is being evaluated separately? I’d expected this to result:

```
(article-pipeline|twitter-pipeline)
```

**Comment by eonpatapon** [comment]
Yes. Each leaf will become a separate cue instance inheriting constrains, fields etc... from parent directories. This is mention here in the doc:  https://cuelang.org/docs/concepts/packages/#instances

The output you have with `cue eval` shows the result of 3 instances: `config/pipelines` (which has no pipelines), `config/pipelines/a` and`config/pipelines/b`.

`a` and `b` inherit the `_monitored_pipelines_label` field which is defined in the parent, but since for each directory a separate instance is created you only see the pipeline of each directory respectively.

If you need to aggregate multiple pipelines in one instance you could:
* define them in a single directory, in multiple files
* define them in separate packages and import them where you need them

**Reply by myitcv** [reply]
Just to add one slight caveat here. The loading mechanisms currently supported by CUE are, to a large extent, informed by the Go project package/module approach, but also the approach/style of Kubernetes configurations (hence the [Kubernetes tutorial](https://github.com/cue-lang/cue/blob/master/doc/tutorial/kubernetes/README.md) and our follow up from [FOSDEM this year](https://github.com/cue-examples/fosdem2022/blob/main/guide/guide_fosdem2022_en.markdown)). 

However, we have seen instances where the file organisation approach this "enforces" simply does not work. The [Vector](https://github.com/vectordotdev/vector) project is one such example, where they have a single package (and hence a single configuration instance) split across multiple files and directories, because that better reflects how they organise and manage the component parts of that single package.

That said, if there are scenarios where you end up feeling "this really does fit/work" please flag them. We are more than happy to help dig into those to uncover whether it's a case of better documentation to help explain the pattern (documentation can always be improved), or whether we really need to consider supporting alternative loading mechanisms/patterns.
